Dotprint: add function to obtain short node names

OliverAh opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #28580

#### Brief description of what is fixed or changed
`sympy.printing.dot.dotprint(...)` creates a string representation
of expressions, to be rendered into graph images by other packages,
e.g. graphviz. This requires unique identifiers of the nodes.
These unique identifiers are by default just the expression's srepr,
which grows with the expression. If this string representation is
too long, graphviz raises an error, and the graph is not rendered.
This commit introduces a switch to create short node names of
the form 'node_0', 'node_1', 'node_2', ...
As counting the node numbers requires a shared counter between
.dotnode and .dotedges, a kwarg nodenamefunc is introduced.
Users might also use it to introduce different functions, but this
is only a side effect. .dotnode and .dotedges can also be
called without .dotprint, for them the default nodenamefunc
is still .purestr from before this PR, because otherwise
shared counters had to be instantiated, or additional checks
would be required.

Example:
In [1]: import sympy
In [2]: from sympy.abc import x
In [3]: import graphviz

In [4]: expr = sympy.sin(x)
In [5]: for i in range(8):
........... expr = sympy.Derivative(expr, x) + sympy.sin(expr)

In [6]: string_dotprint = sympy.printing.dot.dotprint(expr,
shortnames=True)
In [7]: graph_source = graphviz.Source(string_dotprint)
In [8]: graph_source.render(filename='mwe_breaking_sympy_graphviz_large', format='svg')

Without the new function for short node names, used in line 6, graphviz would raise an error in
line 8 and not render the graph.

#### Other comments
Note: the node names are NOT the node labels, which are rendered into the graph. These can still be adjusted
by `labelfunc=Callable` as before. This PR does not interfere with any other functionality.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

<!-- BEGIN RELEASE NOTES -->

* printing
  * Added a new function for `dotprint` to produce short node names of the form 'node_0', 'node_1', ... instead of the expression's `srepr`. 
    This allows to render also large expressions into graphs, which was previously not possible using `dotprint`.

<!-- END RELEASE NOTES -->

Commit:
dotprint: added switch to use short node names

closes #28580.
sympy.printing.dot.dotprint(...) creates a string representation
of expressions, to be rendered into graph images by other packages,
e.g. graphviz. This requires unique identifiers of the nodes.
These unique identifiers are by default just the expression's srepr,
which grows with the expression. If this string representation is
too long, graphviz raises an error, and the graph is not rendered.
This commit introduces a switch to create short node names of
the form 'node_0', 'node_1', 'node_2', ...

Example:
In [1]: import sympy
In [2]: from sympy.abc import x
In [3]: import graphviz

In [4]: expr = sympy.sin(x)
In [5]: for i in range(8):
........... expr = sympy.Derivative(expr, x) + sympy.sin(expr)

In [6]: string_dotprint = sympy.printing.dot.dotprint(expr,
shortnames=True)
In [7]: graph_source = graphviz.Source(string_dotprint)
In [8]: graph_source.render(filename='mwe_breaking_sympy_graphviz_large', format='svg')

Without 'shortnames=True' in line 6, graphviz would raise an error in
line 8 and not render the graph. The default is 'False', which is the
same behaviour as before this commit.

Commit:
author: add OliverAh to .mailmap

Commit:
author: add OliverAh to .mailmap

OliverAh commented on sympy/printing/dot.py:
```diff
@@ -185,7 +250,7 @@ def dotedges(expr, atom=lambda x: not isinstance(x, Basic), pos=(), repeat=True)
 def dotprint(expr,
     styles=default_styles, atom=lambda x: not isinstance(x, Basic),
-    maxdepth=None, repeat=True, labelfunc=str, **kwargs):
+    maxdepth=None, repeat=True, labelfunc=str, shortnodenames=False, **kwargs):
```
My reasoning for the switch is, that `dotprint` does not immediately render the graphs, but only creates this intermediate string, which would then be fed into e.g. graphviz. But since the current implementation with the srepr is quite specific, and allows to easily reconstruct the whole expression, others might rely on this wholistic representation of the expression.
My only usecase for the dot-string is to forward it to graphviz, but in general there could be more.

oscarbenjamin replied:
Maybe others have views on this but I would rather just change it and not have any config option.

OliverAh replied:
Since there seems to be no additional opinion, I will remove the switch and update the PR.

Commit:
author: add OliverAh to .mailmap

Commit:
dotprint: added new function for short node names

closes #28580.
sympy.printing.dot.dotprint(...) creates a string representation
of expressions, to be rendered into graph images by other packages,
e.g. graphviz. This requires unique identifiers of the nodes.
These unique identifiers were just the expression's srepr,
which grows with the expression. If this string representation is
too long, graphviz raises an error, and the graph is not rendered.
This commit introduces a new function to create short node names of
the form 'node_0', 'node_1', 'node_2', ... by default.
As counting the node numbers requires a shared counter between
.dotnode and .dotedges, a kwarg nodenamefunc is introduced.
Users might also use it to introduce different functions, but this
is only a side effect. .dotnode and .dotedges can also be
called without .dotprint, for them the default nodenamefunc
is still .purestr from before this PR, because otherwise
shared counters had to be instantiated, or additional checks
would be required.

Example:
In [1]: import sympy
In [2]: from sympy.abc import x
In [3]: import graphviz

In [4]: expr = sympy.sin(x)
In [5]: for i in range(8):
........... expr = sympy.Derivative(expr, x) + sympy.sin(expr)

In [6]: string_dotprint = sympy.printing.dot.dotprint(expr,
shortnames=True)
In [7]: graph_source = graphviz.Source(string_dotprint)
In [8]: graph_source.render(filename='mwe_breaking_sympy_graphviz_large', format='svg')

Without the new short node names, generated in line 6, graphviz would raise an error in
line 8 and not render the graph.

Commit:
adapted doctest for new short node names

Commit:
completely avoid options for nodename generation when calling dotprint
